clang/lib/3C/DeclRewriter.cpp

```diff
@@ -450,9 +511,13 @@ void DeclRewriter::rewriteMultiDecl(MultiDeclInfo &MDI, RSet &ToRewrite) {
           getDeclSourceRangeWithAnnotations(DL, /*IncludeInitializer=*/true);
       SourceRange Comma = getNextComma(SkipSR.getEnd());
       rewriteSourceRange(R, Comma, ";\n");
-      // Offset by one to skip past what we've just added so it isn't
-      // overwritten.
-      PrevEnd = Comma.getEnd().getLocWithOffset(1);
+      // Offset to skip past what we've just added so it isn't overwritten.
+      // If a duplicate declaration was emitted for as part of rewriting a
+      // declaration, then the offset must be 2 to skip past that as well.
+      unsigned int Offset = 1;
+      if (Replacement && !Replacement->getSupplementaryDecls().empty())
+        Offset = 2;
```

Comment:
As John and I previously discussed, this is weird.  Source locations are counted in characters in the original file; a rewriter insertion by itself doesn't count as a character.  So this code is effectively preserving one extra character after the comma.  Example:
```c
#include <stdlib.h>

void test4() {
  int *a = malloc(10*sizeof(int)), b;
  int *c = malloc(10*sizeof(int)),d;
  a++; c++;
}
```
Output:
```c
#include <stdlib.h>

void test4() {
  _Array_ptr<int> __3c_tmp_a : count(10) = malloc<int>(10*sizeof(int));

_Array_ptr<int> a : bounds(__3c_tmp_a, __3c_tmp_a + 10) = __3c_tmp_a;
 int b;  // The space after the comma ends up at the beginning of this line.
  _Array_ptr<int> __3c_tmp_c : count(10) = malloc<int>(10*sizeof(int));

_Array_ptr<int> c : bounds(__3c_tmp_c, __3c_tmp_c + 10) = __3c_tmp_c;
dint d;  // The `d` after the comma gets duplicated at the beginning of this line: oops.
  a++; c++;
}
```
This is not ideal; we should find a better solution if we can.  But if I just remove the offset adjustment, we get garbage output; no doubt that was why John added it in the first place.

After further investigation, I realized that the garbage is just due to a bug in `Rewriter::ReplaceText` that I noticed back on July 15 while I was reading the code to learn about the design possibilities for struct de-nesting (see #531).  (Once in a while, I get something useful out of extra time I spend learning about something!)  If I work around the bug, then I can remove the offset adjustment and the regression tests in this PR still pass: see [the `range_bounds.matt_review_offset` branch](https://github.com/correctcomputation/checkedc-clang/compare/range_bounds...range_bounds.matt_rewrite_offset).  John, if you're happy with that approach, please merge it into this PR.  Also please consider adding a regression test that the `d` doesn't get duplicated in the `int *c = malloc(10*sizeof(int)),d;` case.

Reply:
As agreed at today's meeting, I wrote [the regression test](https://github.com/correctcomputation/checkedc-clang/compare/range_bounds...range_bounds.no_space_test).  I checked that it fails on the old code as shown above, so it is testing the right thing.